chromium_src/third_party/blink/renderer/modules/battery/battery_manager.cc

```diff
@@ -0,0 +1,90 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "third_party/blink/renderer/modules/battery/battery_manager.h"
+
+#include "third_party/blink/renderer/core/dom/document.h"
+#include "third_party/blink/renderer/core/dom/dom_exception.h"
+#include "third_party/blink/renderer/core/dom/events/event.h"
+#include "third_party/blink/renderer/core/execution_context/execution_context.h"
+#include "third_party/blink/renderer/modules/battery/battery_dispatcher.h"
+#include "third_party/blink/renderer/platform/wtf/assertions.h"
+
+namespace blink {
+
+  BatteryManager::~BatteryManager() = default;
+
+  BatteryManager::BatteryManager(ExecutionContext* context)
+  : PausableObject(context), PlatformEventController(ToDocument(context)) {}
+
+  BatteryManager* BatteryManager::Create(ExecutionContext* context) {
+    BatteryManager* battery_manager = new BatteryManager(context);
+    battery_manager->PauseIfNeeded();
+    return battery_manager;
+  }
+
+  ScriptPromise BatteryManager::StartRequest(ScriptState* script_state) {
+    if (!battery_property_) {
+      battery_property_ = new BatteryProperty(
+            ExecutionContext::From(script_state), this, BatteryProperty::kReady);
+      battery_property_->Resolve(this);
+      }
+      return battery_property_->Promise(script_state->World());
+  }
+
+  bool BatteryManager::charging() {
+    return true;
+  }
+
+  double BatteryManager::chargingTime() {
+    return 0;
+  }
+
+  double BatteryManager::dischargingTime() {
+    return std::numeric_limits<double>::infinity();
+  }
+
+  double BatteryManager::level() {
+    return 1.0;
+  }
+
+  void BatteryManager::Trace(blink::Visitor* visitor) {
+    visitor->Trace(battery_property_);
+    PlatformEventController::Trace(visitor);
+    EventTargetWithInlineData::Trace(visitor);
+    PausableObject::Trace(visitor);
+  }
+
+  bool BatteryManager::HasPendingActivity() const {
+    return false;
+  }
+
+  void BatteryManager::Pause() { 
+    return;
+  }
+
+  void BatteryManager::Unpause() { 
+    return;
+  }
+
+  void BatteryManager::UnregisterWithDispatcher() {
+    return;
```

Comment:
upstream sets this to null here and I think that might be necessary for GC of the `Member` property

Reply:
https://github.com/brave/brave-core/pull/567/files#diff-70de47d152a190708103966693ce09e9R80

Reply:
I should have been more specific that `this` is `battery_property_`

Reply:
sorry, wrong method. Happens in ContextDestroyed below which is now correct